### PR TITLE
Don't accept 0000:00:00 00:00:00 as a valid date

### PR DESCRIPTION
**Pull request by kamil4**
Fixes https://github.com/LycheeOrg/Lychee/issues/1200

I probably missed some cases but I only had two samples to play with...

### CODE REVIEW

**Comment by kamil4** [comment]
Looks like CI is failing with PHP 8.0 because imagick is not configured?

**Review by nagmat84 — APPROVED** [review]
I haven't tested it, but it looks reasonable to me. I saw that you compare the various different type of dates with slightly different strings, i.e.

 * `'0000:00:00 00:00:00'`
 * `'0000:00:00 00:00:00Z'`
 * `'0000:00:00 00:00:00.000000Z'`

Have you checked the specification or did you inspect some sample files? I am only curious how you came up with these different formats for the different attributes.

**Comment by kamil4** [comment]
Yeah, I didn't like that part about my PR. Basically, I ran exiftool, ffprobe, etc. on a variety of files and included the date formats I saw in the patch. But of course these things are not set in stone -- they could add microseconds or time zone in the next version and then it wouldn't match.

So I just pushed a commit that instead uses a single regular expression that matches all the date formats I encountered -- hopefully that will be less fragile...

File: lib/PHPExif/Mapper/Exiftool.php [hunk 1]
```diff
@@ -229,7 +229,8 @@ public function mapRawData(array $data)
                 case self::DATETIMEORIGINAL_WMV:
                     // DATETIMEORIGINAL_APPLE contains data on timezone
                     // only set value if DATETIMEORIGINAL_APPLE has not been used
-                    if (!isset($mappedData[Exif::CREATION_DATE])) {
+                    if (!isset($mappedData[Exif::CREATION_DATE])
+                            && preg_match('/^0000:00:00.00:00:00/', $value) === 0) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/Exiftool.php:233]:
```suggestion
                            && preg_match('/^0000[:-]00[:-]00.00:00:00/', $value) === 0) {
```

File: lib/PHPExif/Mapper/Exiftool.php [hunk 2]
```diff
@@ -250,6 +251,9 @@ public function mapRawData(array $data)
 
                     break;
                 case self::DATETIMEORIGINAL_APPLE:
+                    if (preg_match('/^0000:00:00.00:00:00/', $value) === 1) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/Exiftool.php:254]:
```suggestion
                    if (preg_match('/^0000[:-]00[:-]00.00:00:00/', $value) === 1) {
```

File: lib/PHPExif/Mapper/FFprobe.php [hunk 3]
```diff
@@ -120,7 +120,8 @@ public function mapRawData(array $data)
                 case self::DATETIMEORIGINAL:
                     // QUICKTIME_DATE contains data on timezone
                     // only set value if QUICKTIME_DATE has not been used
-                    if (!isset($mappedData[Exif::CREATION_DATE])) {
+                    if (!isset($mappedData[Exif::CREATION_DATE])
+                            && preg_match('/^0000:00:00.00:00:00/', $value) === 0) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/FFprobe.php:124]:
```suggestion
                            && preg_match('/^0000[-:]00[-:]00.00:00:00/', $value) === 0) {
```

File: lib/PHPExif/Mapper/FFprobe.php [hunk 4]
```diff
@@ -135,6 +136,9 @@ public function mapRawData(array $data)
 
                     break;
                 case self::QUICKTIME_DATE:
+                    if (preg_match('/^0000:00:00.00:00:00/', $value) === 1) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/FFprobe.php:139]:
```suggestion
                    if (preg_match('/^0000[-:]00[-:]00.00:00:00/', $value) === 1) {
```

File: lib/PHPExif/Mapper/ImageMagick.php [hunk 5]
```diff
@@ -114,7 +114,8 @@ public function mapRawData(array $data)
                     $value = sprintf('f/%01.1f', $this->normalizeComponent($value));
                     break;
                 case self::CREATION_DATE:
-                    if (!isset($mappedData[Exif::CREATION_DATE])) {
+                    if (!isset($mappedData[Exif::CREATION_DATE])
+                            && preg_match('/^0000:00:00.00:00:00/', $value) === 0) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/ImageMagick.php:118]:
```suggestion
                            && preg_match('/^0000[-:]00[-:]00.00:00:00/', $value) === 0) {
```

File: lib/PHPExif/Mapper/ImageMagick.php [hunk 6]
```diff
@@ -125,6 +126,9 @@ public function mapRawData(array $data)
                     }
                     break;
                 case self::DATETIMEORIGINAL:
+                    if (preg_match('/^0000:00:00.00:00:00/', $value) === 1) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/ImageMagick.php:129]:
```suggestion
                    if (preg_match('/^0000[-:]00[-:]00.00:00:00/', $value) === 1) {
```

File: lib/PHPExif/Mapper/Native.php [hunk 7]
```diff
@@ -178,6 +178,9 @@ public function mapRawData(array $data)
             // manipulate the value if necessary
             switch ($field) {
                 case self::DATETIMEORIGINAL:
+                    if (preg_match('/^0000:00:00.00:00:00/', $value) === 1) {
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/Native.php:181]:
```suggestion
                    if (preg_match('/^0000[-:]00[-:]00.00:00:00/', $value) === 1) {
```

**Review by nagmat84 — COMMENTED** [review]
I expanded the regex. Once, I also wanted to solve the problem and I had already dived into the specs. At least the specs allow to also use a hyphen as a separator between the date components, so I added it.

**Pull request merged**